src/Npgsql/NpgsqlCommand.cs

```diff
@@ -418,7 +584,7 @@ public override void Prepare()
                 if (!needToPrepare)
                     return;
 
-                var sendTask = SendPrepare(false);
+                var sendTask = SendParseDescribe(false, true);
```

Comment:
I think you want to remove SendParseDescribe() and call SendPrepare here, no?

Reply:
No!
`SendPrepare()` is no longer working on its own.
It is now a pure child of `SendParseDescribe()` and has to be called via it (as well as `SendDeriveParameters()` has).
It misses quite some functionality that is common for both child methods.
In practice both methods could be inlined with no effort but this would obfuscate the code a little more and make the SendParseDescribe() more complex again.

Reply:
OK, I understand what you did. I can't say I love having `SendPrepare()` and `SendDeriveParameters()` being called only from `SendParseDescribe()`, whose only purpose is some pretty minimal pre- and post-sending boilerplate code. Note that there's also `SendExecute()`, `SendClose()` etc. which also contain the pre- and post-sending code, so it's not clear why only prepare/derive are wrapped by something like `SendParseDescribe()`. Note also that refactoring out common code when using async methods isn't exactly "free" like regular methods, since async methods have some overhead (i.e. the state machine) - although in this case it's probably a bit marginal.

I'd personally prefer to have simple `SendPrepare()` and `SendDeriveParameters()`, which contain all the necessary logic. Yeah, some boilerplate would be duplicated but I don't think it's a big deal and it would be much more readable. Another option would be to at least implement `SendPrepare()` and `SendDeriveParameters()` as C# 7 local methods - this would at least make it very clear they're only to be called from `SendParseDescribe()`. In this case I'd rename `SendParseDescribe()` to something like `SendPrepareOrDerive()`, since I prefer naming methods based on what they do and not by how they do it or how they're implemented (i.e. nobody outside Npgsql/PostgreSQL knows what Parse and Describe are).

PS Nope, haven't read Clean Code :)

Reply:
> Note also that refactoring out common code when using async methods isn't exactly "free" like regular methods, since async methods have some overhead (i.e. the state machine) - although in this case it's probably a bit marginal.

Thanks for pointing this out. I tend to forget about that. To be honest, async await still doesn't feel natural to me. I always have to think twice about possible side effects of what I'm doing.
In my opinion cases like this are a wonderful case, where *c#-compiler*-inlining would make sense. The *jit-compiler* would obviously have a hard time to inline the whole state machine, but it should be a pretty easy job for the compiler to figure out that the generated IL (and in this case probably also the resulting machine code) is reduced quite a bit.

I thank you also for bringing me back to the pragmatic path. I tend to over-engineer when trying to write dry code, but there shouldn't be a problem with a few lines of duplicated boilerplate code.

I've squashed the latest commit so that hopefully everything should be ready to merge now.